op-node/rollup/driver/sequencer.go

```diff
@@ -12,55 +12,293 @@ import (
+		l1OriginNumber := l2Head.L1Origin.Number
+		batch.l1Origin, err = d.l1OriginSelector.FindL1OriginByNumber(ctx, l1OriginNumber)
+		if err != nil {
+			d.log.Error("Error finding L1 origin with number", l1OriginNumber, "err", err)
+			return err
+		}
+		d.espressoBatch = batch
+		return nil
+	}
+	// 2) Espresso skipped an L1 block.
+	if batch.jst.FirstBlock.L1Block.Number > l2Head.L1Origin.Number+1 {
+		// Produce an empty batch that advances the L1 origin by 1, so we can catch up to Espresso.
+		l1OriginNumber := l2Head.L1Origin.Number + 1
+		batch.l1Origin, err = d.l1OriginSelector.FindL1OriginByNumber(ctx, l1OriginNumber)
+		if err != nil {
+			d.log.Error("Error finding L1 origin with number", l1OriginNumber, "err", err)
+			return err
+		}
+		d.espressoBatch = batch
+		return nil
+	}
+
+	// Fetch the L1 origin determined by the first Espresso block.
+	l1OriginNumber := batch.jst.FirstBlock.L1Block.Number
+	batch.l1Origin, err = d.l1OriginSelector.FindL1OriginByNumber(ctx, l1OriginNumber)
+	if err != nil {
+		d.log.Error("Error finding L1 origin with number", l1OriginNumber, "err", err)
+		return err
+	}
+
+	// Check for one more case where the L1 origin is ineligible: if it is too old, we produce an
+	// empty batch that advances the L1 origin by 1.
+	if batch.l1Origin.Time+d.config.MaxSequencerDrift < windowStart {
+		l1OriginNumber = l2Head.L1Origin.Number + 1
+		batch.l1Origin, err = d.l1OriginSelector.FindL1OriginByNumber(ctx, l1OriginNumber)
+		if err != nil {
+			d.log.Error("Error finding L1 origin with number", l1OriginNumber, "err", err)
+			return err
+		}
+		d.espressoBatch = batch
+		return nil
+	}
+
+	// If we didn't hit any of the edge cases above, we are eligible to include transactions
+	// produced by Espresso in this batch.
+	batch.jst.Payload = &eth.L2BatchPayloadJustification{
+		// We have not included any blocks yet, so the "last block" (the block to start fetching new
+		// blocks after) is one _before_ the start of the window. `updateEspressoBatch` will update
+		// this field as the blocks we have fetched get inserted into the batch.
+		LastBlock: batch.jst.PrevBatchLastBlock,
+		// `NextBatchFirstBlock` must always be one block after `LastBlock`. `updateEspressoBatch`
+		// will also keep this field in sync.
+		NextBatchFirstBlock: batch.jst.FirstBlock,
+		// We haven't added any proofs yet.
+		NmtProofs: nil,
+	}
+	d.espressoBatch = batch
+	return d.updateEspressoBatch(ctx, blocks)
+}
+
+// updateEspressoBatch appends the transactions contained in the Espresso blocks denoted by
+// `newHeaders` to the current in-progress batch. If the batch is complete (Espresso has sequenced
+// at least one block with a timestamp beyond the end of the current sequencing window) it will set
+// the `complete` flag.
+func (d *Sequencer) updateEspressoBatch(ctx context.Context, newHeaders []espresso.Header) error {
+	batch := d.espressoBatch
+	for i := range newHeaders {
+		if newHeaders[i].Timestamp >= batch.windowEnd {
```

Comment:
Maybe log some sort of warning if ```i != newHeaders.length? ``` This would indicate a weird bug. We could also consider enforcing that newHeaders is within the window range and have explicit startHeader and endHeader fields so that instead of iterating over newHeaders until we leave the window range, we check endHeader.

Reply:
Yeah this would be much cleaner